include/itkAttenuationImageFilter.hxx

```diff
@@ -0,0 +1,225 @@
+{
+  // Verify inputs
+  if (this->GetMaskImage() == nullptr)
+  {
+    itkExceptionMacro("Filter requires a mask image for inclusion estimates!");
+  }
+
+  if (this->GetDirection() > ImageDimension)
+  {
+    itkExceptionMacro("Scan line direction must be a valid image dimension!");
+  }
+
+  if (this->GetSamplingFrequencyMHz() < itk::Math::eps)
+  {
+    itkExceptionMacro("RF sampling frequency was not set!");
+  }
+
+  Superclass::BeforeThreadedGenerateData();
+
+  // Initialize metric image
+  this->GetOutput()->Allocate();
+  this->GetOutput()->FillBuffer(0.0f);
+}
+
+template <typename TInputImage, typename TOutputImage, typename TMaskImage>
+void
+AttenuationImageFilter<TInputImage, TOutputImage, TMaskImage>::AfterThreadedGenerateData()
+{
+  Superclass::AfterThreadedGenerateData();
+}
+
+template <typename TInputImage, typename TOutputImage, typename TMaskImage>
+void
+AttenuationImageFilter<TInputImage, TOutputImage, TMaskImage>::DynamicThreadedGenerateData(
+  const OutputRegionType & regionForThread)
+{
+  if (regionForThread.GetNumberOfPixels() == 0)
+  {
+    return;
+  }
+
+  const MaskImageType *  maskImage = this->GetMaskImage();
+  const InputImageType * input = this->GetInput();
+  OutputImageType *      output = this->GetOutput();
+
+  ImageLinearConstIteratorWithIndex<TInputImage> it(input, regionForThread);
+  it.SetDirection(m_Direction);
+  it.GoToBegin();
+
+  unsigned int   inclusionLength;
+  InputIndexType start, end;
+
+  const float scanStepMM = input->GetSpacing()[m_Direction];
+
+  // do the work
+  while (!it.IsAtEnd())
+  {
+    inclusionLength = 0;
+    while (!it.IsAtEndOfLine())
+    {
+      // Advance until an inclusion is found
+      InputIndexType index = it.GetIndex();
+      if (maskImage->GetPixel(index) == m_LabelValue && inclusionLength == 0)
+      {
+        // Step into inclusion
+        start = it.GetIndex();
+      }
+      else if (maskImage->GetPixel(index) != m_LabelValue && inclusionLength > 0)
+      {
+        // Step out of inclusion
+        end = it.GetIndex();
+
+        // Adjust for pixel padding
+        start[m_Direction] += m_PadPixelUpperBounds;
+        end[m_Direction] -= m_PadPixelLowerBounds;
+
+        // Adjust for spatial padding
+        start[m_Direction] += static_cast<unsigned int>(m_PadSpatialUpperBoundsMM / scanStepMM);
+        end[m_Direction] -= static_cast<unsigned int>(m_PadSpatialLowerBoundsMM / scanStepMM);
+
+        // Determine spatial distance between two scanline points and then
+        // adjust for spatial padding
+        unsigned int pixelDistance = end[m_Direction] - start[m_Direction];
+        float        distanceMM = pixelDistance * scanStepMM;
+
+        // Estimate attenuation for each inclusion pixel
+        // with respect to the last pixel in the inclusion
+        while (start[m_Direction] < end[m_Direction])
+        {
+          float estimatedAttenuation = ComputeAttenuation(end, start);
```

Comment:
```suggestion
          InputIndexType target = end;
          if (m_FixedEstimationDepth!=0 && end[m_Direction] - start[m_Direction] > m_FixedEstimationDepth)
          {
            target[m_Direction] = start[m_Direction] + fixedDistance;
          }
          float estimatedAttenuation = ComputeAttenuation(target, start);
```

Reply:
Is `fixedDistance` here intended as a pixel distance approximating the spatial distance from `m_FixedEstimationDepth`? And is it valid to possibly discard pixels at the end of the inclusion based on `m_FixedEstimationDepth`?

Reply:
> pixel distance approximating the spatial distance

Yes.

> discard pixels at the end of the inclusion based on m_FixedEstimationDepth

For typical use-case, that would seriously erode the inclusion mask. Better not do it, or maybe do it as an option.

Reply:
> For typical use-case, that would seriously erode the inclusion mask

Understood. I've added an inline comment to clarify the procedure.